core/server/worker/src/main/java/alluxio/worker/netty/DataServerBlockReadHandler.java

```diff
@@ -84,62 +114,108 @@ public void close() throws IOException {
    * @param blockWorker the block worker
    * @param fileTransferType the file transfer type
    */
-  DataServerBlockReadHandler(ExecutorService executorService, BlockWorker blockWorker,
+  public DataServerBlockReadHandler(ExecutorService executorService, BlockWorker blockWorker,
       FileTransferType fileTransferType) {
     super(executorService);
     mWorker = blockWorker;
     mTransferType = fileTransferType;
   }
 
   @Override
-  protected boolean acceptMessage(Object object) {
-    if (!super.acceptMessage(object)) {
-      return false;
-    }
-    Protocol.ReadRequest request = ((RPCProtoMessage) object).getMessage().asReadRequest();
-    return request.getType() == Protocol.RequestType.ALLUXIO_BLOCK;
-  }
-
-  @Override
   protected void initializeRequest(Protocol.ReadRequest request) throws Exception {
     mRequest = new BlockReadRequestInternal(request);
   }
 
   @Override
-  protected DataBuffer getDataBuffer(Channel channel, long offset, int len) throws IOException {
-    LocalFileBlockReader blockReader =
-        (LocalFileBlockReader) ((BlockReadRequestInternal) mRequest).mBlockReader;
-    Preconditions.checkArgument(blockReader.getChannel() instanceof FileChannel,
-        "Only FileChannel is supported!");
-    switch (mTransferType) {
-      case MAPPED:
-        ByteBuf buf = channel.alloc().buffer(len, len);
+  protected DataBuffer getDataBuffer(Channel channel, long offset, int len) throws Exception {
+    openBlock(channel);
+    BlockReader blockReader = ((BlockReadRequestInternal) mRequest).mBlockReader;
+
+    if (mTransferType == TRANSFER && !(blockReader instanceof UnderFileSystemBlockReader)) {
+      return new DataFileChannel(new File(((LocalFileBlockReader) blockReader).getFilePath()),
+          offset, len);
+    } else {
+      ByteBuf buf = channel.alloc().buffer(len, len);
+      try {
+        while (buf.writableBytes() > 0 && blockReader.transferTo(buf) != -1) {
+        }
+        return new DataNettyBufferV2(buf);
+      } catch (Throwable e) {
+        buf.release();
+        throw e;
+      }
+    }
+  }
+
+  /**
+   * Opens the UFS block if it is not open.
+   *
+   * @param channel the netty channel
+   * @throws Exception if it fails to open the UFS block
+   */
+  private void openBlock(Channel channel) throws Exception {
+    BlockReadRequestInternal request = (BlockReadRequestInternal) mRequest;
+    if (request.mBlockReader != null) {
+      return;
+    }
+
+    int retryInterval = Constants.SECOND_MS;
+    RetryPolicy retryPolicy = new TimeoutRetry(UFS_BLOCK_OPEN_TIMEOUT_MS, retryInterval);
+
+    do {
+      long lockId;
+      if (request.mOpenUfsBlockOptions != null) {
```

Comment:
Consider adding a `has` method instead of doing a null check to make this easier to maintain?